src/bmp_matcher.rs

```diff
@@ -0,0 +1,305 @@
+
+#[derive(Debug, Clone, Default)]
+pub struct BmpMatcher
+{
+	index: Option<usize>,
+	serial: Option<String>,
+	port: Option<PortId>,
+}
+
+enum MatchResult
+{
+	NoMatch(DeviceInfo),
+	Found(BmpDevice),
+	Error(Report),
+}
+
+impl BmpMatcher
+{
+	pub fn new() -> Self
+	{
+		Default::default()
+	}
+
+	pub fn new_with_port(port: PortId) -> Self
+	{
+		BmpMatcher {
+			port: Some(port),
+			..Default::default()
+		}
+	}
+
+	pub fn from_params<Params>(params: &Params) -> Self
+	where
+		Params: BmpParams,
+	{
+		Self::new()
+			.index(params.index())
+			.serial(params.serial_number())
+			.port(None)
+	}
+
+	/// Set the index to match against.
+	#[must_use]
+	pub fn index(mut self, idx: Option<usize>) -> Self
+	{
+		self.index = idx;
+		self
+	}
+
+	/// Set the serial number to match against.
+	#[must_use]
+	pub fn serial<'s, IntoOptStrT>(mut self, serial: IntoOptStrT) -> Self
+	where
+		IntoOptStrT: Into<Option<&'s str>>,
+	{
+		self.serial = serial.into().map(|s| s.to_string());
+		self
+	}
+
+	/// Set the port path to match against.
+	#[must_use]
+	pub fn port(mut self, port: Option<PortId>) -> Self
+	{
+		self.port = port;
+		self
+	}
+
+	/// Get any index previously set with `.index()`.
+	pub fn get_index(&self) -> Option<usize>
+	{
+		self.index
+	}
+
+	/// Get any serial number previously set with `.serial()`.
+	pub fn get_serial(&self) -> Option<&str>
+	{
+		self.serial.as_deref()
+	}
+
+	/// Get any port path previously set with `.port()`.
+	pub fn get_port(&self) -> Option<PortId>
+	{
+		self.port.clone()
+	}
+
+	/// Find all connected Black Magic Probe devices that match from the command-line criteria.
+	///
+	/// This uses the `serial_number`, `index`, and `port` values from `matches`, treating any that
+	/// were not provided as always matching.
+	///
+	/// This function returns all found devices and all errors that occurred during the search.
+	/// This is so errors are not hidden, but also do not prevent matching devices from being found.
+	/// However, if the length of the error `Vec` is not 0, you should consider the results
+	/// potentially incomplete.
+	///
+	/// The `index` matcher *includes* devices that errored when attempting to match them.
+	pub fn find_matching_probes(&self) -> BmpMatchResults
+	{
+		let mut results = BmpMatchResults {
+			found: Vec::new(),
+			filtered_out: Vec::new(),
+			errors: Vec::new(),
+		};
+
+		let devices = match list_devices() {
+			Ok(d) => d,
+			Err(e) => {
+				results.errors.push(e.into());
+				return results;
+			},
+		};
+
+		// Filter out devices that don't match the Black Magic Probe's vid/pid in the first place.
+		let devices = devices.filter(|dev| {
+			let vid = dev.vendor_id();
+			let pid = dev.product_id();
+			BmpPlatform::from_vid_pid(Vid(vid), Pid(pid)).is_some()
+		});
+
+		devices
+			.enumerate()
+			.map(|(index, device_info)| self.matching_probe(index, device_info))
+			.collect()
+	}
+
+	fn matching_probe(&self, index: usize, device_info: DeviceInfo) -> MatchResult
```

Comment:
This function makes sense, just please include a `///` comment documenting why it exists and what it's aiming to do (the function's intent) please.